libraries/AP_MSP/AP_MSP_Telem_Backend.cpp

```diff
@@ -0,0 +1,801 @@
+                uint8_t end_position = start_position + MSP_TXT_VISIBLE_CHARS;
+
+                //ensure array boundaries
+                start_position = MIN(start_position, int(sizeof(buffer)-1));
+                end_position = MIN(end_position, int(sizeof(buffer)-1));
+
+                //trim invisible part
+                buffer[end_position] = 0;
+            }
+
+            sbuf_write_data(dst, buffer + start_position, strlen(buffer + start_position));  // max MSP_TXT_VISIBLE_CHARS chars general text...
+        } else {
+            bool wind_en = false;
+#if OSD_ENABLED
+            wind_en = AP::osd()->screen[0].wind.enabled;
+#endif
+            if (wind_en) {
+                telem_update_wind(_msp.telemetry_info);
+            } else {
+                telem_update_flight_mode(_msp.telemetry_info);
+            }
+            sbuf_write_data(dst, _msp.telemetry_info.flight_mode_str, MSP_TXT_BUFFER_SIZE);  // rendered as up to MSP_TXT_VISIBLE_CHARS chars with UTF8 support
+        }
+    }
+    return true;
+}
+
+bool AP_MSP_Telem_Backend::msp_process_out_status(sbuf_t *dst)
+{
+    uint32_t mode_bitmask = get_osd_flight_mode_bitmask();
+    sbuf_write_u16(dst, 0);                     // task delta time
+    sbuf_write_u16(dst, 0);                     // I2C error count
+    sbuf_write_u16(dst, 0);                     // sensor status
+    sbuf_write_data(dst, &mode_bitmask, 4);     // unconditional part of flags, first 32 bits
+    sbuf_write_u8(dst, 0);
+
+    sbuf_write_u16(dst, constrain_int16(0, 0, 100));  //system load
+    sbuf_write_u16(dst, 0);                     // gyro cycle time
+
+    // Cap BoxModeFlags to 32 bits
+    sbuf_write_u8(dst, 0);
+
+    // Write arming disable flags
+    sbuf_write_u8(dst, 1);
+    sbuf_write_u32(dst, !AP::notify().flags.armed);
+
+    // Extra flags
+    sbuf_write_u8(dst, 0);
+    return true;
+}
+
+bool AP_MSP_Telem_Backend::msp_process_out_osd_config(sbuf_t *dst)
+{
+    sbuf_write_u8(dst, OSD_FLAGS_OSD_FEATURE);                      // flags
+    sbuf_write_u8(dst, 0);                                          // video system
+    // Configuration
+    uint8_t units = OSD_UNIT_METRIC;
+#if OSD_ENABLED
+    units = AP::osd()->units == AP_OSD::UNITS_METRIC ? OSD_UNIT_METRIC : OSD_UNIT_IMPERIAL;
+#endif
+
+    sbuf_write_u8(dst, units);                                 // units
+    // Alarms
+    sbuf_write_u8(dst, _msp.osd_config.rssi_alarm);                 // rssi alarm
+    sbuf_write_u16(dst, _msp.osd_config.cap_alarm);                 // capacity alarm
+    // Reuse old timer alarm (U16) as OSD_ITEM_COUNT
+    sbuf_write_u8(dst, 0);
+    sbuf_write_u8(dst, OSD_ITEM_COUNT);                             // osd items count
+
+    sbuf_write_u16(dst, _msp.osd_config.alt_alarm);                 // altitude alarm
+
+    // element position and visibility
+    uint16_t pos = 0;
+    for (uint8_t i = 0; i < OSD_ITEM_COUNT; i++) {
+        pos = MSP_OSD_POS_HIDDEN;
```

Comment:
ok, but with msp osds we always send all telemetry (with the other messages) even if items are hidden, this particular msp message only sends the item positions and expects to have positions for all items since items are ordered and you cannot skip any of them